Wrap Go 1.13 error functions

palsivertsen opened the pull request:
## Summary
Add assertions for `error.Is` and `error.As` functions.

## Changes
Example implementation of ErrorAs and ErrorIs

## Motivation
See #970 

## Related issues
Closes #970

palsivertsen commented:
Anyone know a good way to make changes Go1.13 only? I'm playing around with build tags, but it requires changes to `_codegen`.

boyan-soubachov commented:
> Anyone know a good way to make changes Go1.13 only? I'm playing around with build tags, but it requires changes to `_codegen`.

Thanks for your contribution. We've actually had previous attempts at this (https://github.com/stretchr/testify/pull/884) but we had the same problem of backwards compatibility with Go < 1.13.

Build tags is one way but you'd effectively have to fragment the code-base for pre- and post- Go 1.13 code.

boyan-soubachov commented:
I have removed Go 1.10 from our Travis builds. You should not have issues with this anymore since Go 1.13 is the earliest version we officially support anyways.

palsivertsen commented:
> I have removed Go 1.10 from our Travis builds. You should not have issues with this anymore since Go 1.13 is the earliest version we officially support anyways.

@boyan-soubachov I updated the readme to reflect this.

Ready for review.

palsivertsen commented:
Discovered panic if caller passes nil values. Working on fix.

boyan-soubachov commented on README.md:
```diff
@@ -321,7 +321,7 @@ To update Testify to the latest version, use `go get -u github.com/stretchr/test
 ==================
 
-We support the three major Go versions, which are 1.11, 1.12, and 1.13 at the moment.
+We support the Go versions 1.13 and greater.
```
Thanks for doing this. Do you mind explicitly listing them? (i.e. 1.13 and 1.14).

palsivertsen replied:
Not at all. Changed the wording to match the original, with the exception of only listing two major versions.

boyan-soubachov commented on assert/assertions_test.go:
```diff
@@ -2376,3 +2376,75 @@ func Test_truncatingFormat(t *testing.T) {
+		{nil, nil, true},
+	}
+	for _, tt := range tests {
+		tt := tt
```
What is the purpose of this assignment?

nbaztec replied:
Closure capture in case the tests are run in parallel, I reckon.

palsivertsen replied:
^ this

boyan-soubachov replied:
Of course, forgot about that :D Thank you!

boyan-soubachov commented on assert/assertions_test.go:
```diff
@@ -2376,3 +2376,75 @@ func Test_truncatingFormat(t *testing.T) {
+		{nil, nil, false},
+	}
+	for _, tt := range tests {
+		tt := tt
```
Likewise, as mentioned above. What is the purpose of this assignment?

palsivertsen replied:
If the tests are ran in parallel, then `tt` variable might change before `t.Run` has finished.

I set up and example to demonstrate what I mean here:
https://play.golang.org/p/C3RnCbApuDu

boyan-soubachov commented:
@glesica , @mvdkleijn , would you like to give this a look as well before we merge?

mvdkleijn commented:
Will do, probably tonight

SVilgelm commented on assert/assertion_forward.go:
```diff
@@ -204,6 +204,42 @@ func (a *Assertions) Error(err error, msgAndArgs ...interface{}) bool {
+	if h, ok := a.t.(tHelper); ok {
+		h.Helper()
+	}
+	return ErrorAsf(a.t, err, target, msg, args...)
```
since `ErrorAsf` the function is a wrapper for `ErrorAs`, is it better to use `ErrorAs` here?

palsivertsen replied:
Yes, that's a nice optimization! Implementing it would require changes to the `_codegen` tool which is out of scope for this PR. We could make a separate issue for it?

SVilgelm replied:
Oh, I didn't know that it was generated.

SVilgelm replied:
you are right, it's out of scope of this task.

colega commented:
Can we draft a release with this? 🙏

boyan-soubachov commented:
This will go out in the next release (v1.7.0) at the end of the year. We don’t want to rush so as to give it a chance to soak and for any issues by early adopters to surface.

If you want to explicitly use it, just ‘go get’ using the ‘@master’ version tag.

> On 23 Jul 2020, at 20:40, Oleg Zaytsev <notifications@github.com> wrote:
> 
> ﻿
> Can we draft a release with this? 🙏
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub, or unsubscribe.

mvdkleijn commented:
Sorry, didn't get to this. Life's been a little weird lately. :smile:
